tests/mpmc_channel.rs

```diff
@@ -1413,4 +1433,118 @@ mod if_std {
             assert_eq!(count, 11);
         }
     }
+
+    #[test]
+    fn cancel_send_mid_wait() {
+        let (sender, receiver) = channel::<i32>(3);
+        let (waker, count) = new_count_waker();
+        let cx = &mut Context::from_waker(&waker);
+
+        assert_send!(cx, &sender, 5);
+        assert_send!(cx, &sender, 6);
+        assert_send!(cx, &sender, 7);
+
+        {
+            // Cancel a wait in between other waits
+            // In order to arbitrarily drop a non movable future we have to box and pin it
+            let mut poll1 = Box::pin(sender.send(8));
+            let mut poll2 = Box::pin(sender.send(9));
+            let mut poll3 = Box::pin(sender.send(10));
+            let mut poll4 = Box::pin(sender.send(11));
+            let mut poll5 = Box::pin(sender.send(12));
+
+            assert!(poll1.as_mut().poll(cx).is_pending());
+            assert!(poll2.as_mut().poll(cx).is_pending());
+            assert!(poll3.as_mut().poll(cx).is_pending());
+            assert!(poll4.as_mut().poll(cx).is_pending());
+            assert!(poll5.as_mut().poll(cx).is_pending());
+            assert!(!poll1.is_terminated());
+            assert!(!poll2.is_terminated());
+            assert!(!poll3.is_terminated());
+            assert!(!poll4.is_terminated());
+            assert!(!poll5.is_terminated());
+
+            // Cancel 2 futures. Only the remaining ones should get completed
+            // Safety: We are not using these pins again so this is safe.
+            unsafe { std::pin::Pin::into_inner_unchecked(poll2).cancel() };
```

Comment:
The added tests are interesting. They show that in most places where cancellation makes sense the user would have a `Pin<Future>`  and not a `&mut Future`. Maybe the `cancel()` method should be changed to operate on `self: Pin<&mut Self>` instead of `&mut self`.

Since this will follows however the non-shared version I will merge it. And for a next major version both could be changed to use `Pin`.